### ISSUE TITLE
Interfaces for Developers

### ISSUE THREAD

**Issue by YupItsZac** [delight-im/FreeGeoDB#10]
I came across this project today while looking for something to contribute to. It's cool. 

Have you considered also developing libraries/SDks to enable developers to utilize this data in a quick, easy way? 

If you're open to contributions, I can build some PHP/Python/Javascript stuffs to work with this data. Jus a thought.

**Comment by ocram** [comment]
Thanks!

Definitely open to contributions -- but we may have to discuss what such libraries should be doing exactly, right? Probably some lookup and distance calculations, etc.

Anyway, perhaps the most common use case will not be accessing a large JSON/CSV file from code but querying a database that has been filled with the collections.

**Comment by YupItsZac** [comment]
Yea, naturally. I would use your JSON files to populate a DB, then build a small API around that, along with an SDK. 

As far as the acceptable operations: Lookup, Search, Calculate, Nearest, etc. Would be a nice project for weekends and evenings when Im not too busy.

**Comment by YupItsZac** [comment]
Originally I was thinking about MongoDB to store the data, but I feel like MySQL would be easiest since you've already got SQL imports available. 

I'll build the API based off a framework, probably Symfony2. It''ll be REST. Then the SDKs will interact with that API, to access your data.

**Comment by heri** [comment]
+1

I think Postgresql would be more appropriate. It has http://postgis.net/ to calculate geographical queries

**Comment by heri** [comment]
@mwaclawek @YupItsZac If that helps, I just published a Rails app https://github.com/heri/geo-db-api It was for a customer but he was ok on publishing the work

I have used mysql but ideally this needs Postgresql for distance lookups

**Comment by YupItsZac** [comment]
@heri, nice catch. You're right. I haven't used PostgreSQL in the past, so I'll have to do some digging. Thanks for the heads up!

Side note, you're app looks pretty cool.

**Comment by heri** [comment]
@YupItsZac thanks!

Postgres is pretty cool. You can send in latitude and longitude and then ask what's the nearest object :)

**Comment by YupItsZac** [comment]
@heri I'll look into it. How is it in performance? Have you ever interacted with PostgreSQL through PDO or something similar?

**Comment by heri** [comment]
It's good! It's now my database of choice (vs mysql)

**Comment by ocram** [comment]
@YupItsZac That sounds good!

We'll probably add some list of "SDKs and libraries" here where we'll link your separate repository then. Does that make sense?

MongoDB would definitely be a valid choice because working with JSON will be super easy and MongoDB comes with geospatial support:
- http://docs.mongodb.org/manual/reference/operator/query-geospatial/
- http://docs.mongodb.org/master/core/geospatial-indexes/

Working MySQL imports are available, yes, so you can use MySQL as well. Would be a good choice as well!

@heri PostgreSQL with its PostGIS extension is a perfect choice!

But MySQL has such geospatial support as well:
- https://dev.mysql.com/doc/refman/5.5/en/spatial-analysis-functions.html
- https://dev.mysql.com/doc/refman/5.0/en/creating-spatial-indexes.html

Shall we link your Rails app here?

**Comment by heri** [comment]
@mwaclawek I'm more comfortable with mongodb and postgresql so can work on these. 

Yes if you want you can link to the Rails app! 

Thanks

**Comment by YupItsZac** [comment]
@mwaclawek I was looking at MongoDB for this project. I absolutely love that DB. However, I don't know a lot about PostgreSQL, so I might just use that DB so I can learn about it. I like the idea of linking it!

The API is almost done, and I've linked back to this repo in the documentation. I'll start writing the SDKs sometime this week, and should have something workable by Friday (hopefully). 

I saw in #12 that someone wants to build a frontend for it. Maybe that could be a good test of the API/SDK?

**Comment by ocram** [comment]
@heri All I wanted to say is that, whatever database you're most comfortable with, choose _any_ of PostgreSQL, MySQL and MongoDB -- each one will be sufficient for the task :)

Your project has been added to the tools section: https://github.com/delight-im/FreeGeoDB/blob/4d3f114600d459d88c4e99807b69b27f0eea27d2/TOOLS.md

@YupItsZac Choosing a new database (that is well-suited for the task) for learning is a great idea! Your structure and roadmap sound good :)

Might be a good idea to test your SDK, but it's up to the developer, of course. Maybe they want to store the data locally and access it via JavaScript from the client. We'll see :)

**Comment by YupItsZac** [comment]
The first version of the API is done. It took a lot longer than I anticipated because I had to find an apartment in (and move to) Berlin. It's pretty basic right now, but I'll of course build more endpoints and features as time goes (hopefully with suggestions from you guys!). I've included documentation and code samples in PHP and Python. The SDKs aren't quite done yet, but I plan to tackle them soon and get them on Github. This repo is linked in the docs and emails. 

FreeGeo API: http://freegeo.yupitszac.com
